### DISCUSSION TITLE
enhancement: better tts

### DISCUSSION THREAD

**Discussion by LebToki** [open-webui/open-webui#5589] [General]
Problem: 
The current implementation of Whisper in Open-WebUI uses a limited, robotic voice for all interactions. 
While this is functional, it can be jarring and unnatural, making it difficult for users to engage with the interface.

Solution: 
I would like to see the addition of more voices to Open-WebUI, specifically ones that are less robotic and more natural-sounding. This would improve the overall user experience and make the interface more enjoyable to interact with.

Alternatives considered: 
I've considered using third-party voice libraries or integrating existing voice assistants, but these would require significant modifications to the Open-WebUI codebase. I've also considered using text-to-speech software, but these often lack the emotional expression and nuance of human speech.

Additional context: 
The Whisper implementation in Open-WebUI is a great step forward in providing a more natural-sounding interface, but adding more voices would take it to the next level. This would be especially beneficial for users who rely heavily on voice interfaces for daily tasks, such as seniors or individuals with disabilities.

Some potential voices to consider adding include:

- A softer, more gentle voice for more intimate interactions
- A more assertive, commanding voice for tasks that require attention
- A soothing, calming voice for relaxation and stress relief
- A playful, whimsical voice for entertainment and leisure activities

By adding more voices to Open-WebUI, we can create a more engaging and immersive experience for users, making it easier for them to interact with the interface and achieve their goals.

**Comment by zhewang1-intc** [comment]
echo, i think this is a valuable feature.

**Comment by cheahjs** [comment]
Whisper isn't a text to speech model, it handles speech to text.

Currently there's two options for text to speech on Open WebUI, one is generated locally on your browser using the [Web Speech API](https://developer.mozilla.org/en-US/docs/Web/API/Web_Speech_API), and another by calling out to an OpenAI-compatible text to speech API. The local one uses whatever the browser provides, which is typically the OS's text to speech models. On the OpenAI front, some in the community have deployed [LocalAI](https://localai.io/features/text-to-audio/) or [OpenedAI Speech](https://github.com/matatonic/openedai-speech) to provide self-hosted TTS models.

**Comment by MichaelFomenko** [comment]
Local Speech Models:
- SpeechT5: https://huggingface.co/microsoft/speecht5_tts
- Tools: https://huggingface.co/collections/microsoft/speecht5-650995fc647a3ea442cc6c7b

**Comment by boshk0** [comment]
OpenVoice would be great addition to OpenWebUI!
https://github.com/myshell-ai/OpenVoice

**Reply by sanpatricky** [reply]
Is OpenVoice a OpenAI API compatible TTS tool? Cannot find related introduction in OpenVoice page.

**Reply by rsxdalv** [reply]
No, it's a model, not a tool nor an API. It can have APIs made for it.

**Comment by kevin070982** [comment]
I can only recommend to use the solution provided by [UXVirtual](https://github.com/UXVirtual).
When you run Open WebUI in Docker then take a look here: https://github.com/open-webui/open-webui/issues/126#issuecomment-2078407480.

Easy Steps:
Clone https://github.com/matatonic/openedai-speech
Run `docker compose -f docker-compose.min.yml up` for a minimal docker image with only piper TTs support <1GB (CPU only)
or
Run `docker compose up` for the HD version but requires around 4GB GPU VRAM 
Then add the below lines to the TTS Settings under Audio in the Open WebUI Settings.
```
API Base URL: http://host.docker.internal:8000/v1
API Key: sk-111111111
```

following voices can be changed under Set Voice:
alloy
echo
echo-alt
fable
onyx
nova
shimmer

Enjoy more naturally sounding voice.

**Comment by ErroneousBosch** [comment]
There is a very nice solution in piper that has good language support and really high quality voices:
https://github.com/rhasspy/piper

**Comment by reneil1337** [comment]
I'd love to see one of the suggestions made here integrated. favorably the ability to integrated an entirely onprem solution with easy, more-straight forward cloud services as options/alternatives

**Comment by justinh-rahb** [comment]
> I'd love to see one of the suggestions made here integrated. favorably the ability to integrated an entirely onprem solution with easy, more-straight forward cloud services as options/alternatives

We've for a while supported adding external OpenAI-compatible TTS and STT APIs now.

**Comment by chrishart0** [comment]
Here's my setup for using [openedai-speech](https://github.com/matatonic/openedai-speech) which mocks the openai TTS api to run local models to generate text-to-speech. 

My Repo: https://github.com/chrishart0/open-webui-nvidia-cuda-setup

The important part is I added a openedai-speech container in my [docker-compose.yaml](
https://github.com/chrishart0/open-webui-nvidia-cuda-setup/blob/master/docker-compose.yaml)

```
  openedai-speech:
    image: ghcr.io/matatonic/openedai-speech
    container_name: openedai-speech
    # env_file: .env
    environment:
      - 'TTS_HOME=voices'
      - 'HF_HOME=voices'
      - 'PRELOAD_MODEL=xtts'
    ports:  # No need to expose ports, this is a backend service
      - "8087:8000"
    volumes:
      - ./data/openedai-speech/voices:/app/voices
      - ./data/openedai-speech/config:/app/config
    restart: unless-stopped
```

I populate the needed configs for openedai-speech, which is these two files [pre_process_map.default.yaml](https://github.com/chrishart0/open-webui-nvidia-cuda-setup/blob/master/data/openedai-speech/config/pre_process_map.default.yaml), [voice_to_speaker.default.yaml](https://github.com/chrishart0/open-webui-nvidia-cuda-setup/blob/master/data/openedai-speech/config/voice_to_speaker.default.yaml)

Then I configure the local API in the TTS settings. In my repo I do this via an .env file, but you could just do it through the open web UI GUI too: 
![image](https://github.com/open-webui/open-webui/assets/64378248/26e8ad85-4329-40fe-b972-7c2d0fb07fe3)

If you've got a GPU that supports CUDA you should be able to use my repo. If you have any questions just let me know.

**Comment by PeterHagen** [comment]
For additional information, Piper is also available through the [Wyoming protocol](https://github.com/rhasspy/wyoming-piper), which can provide fast responses. This is implemented for use in Home Assistant. A Docker image is also available. For me Piper runs fine without a GPU. Might be an interesting option, in my opinion.

**Comment by thiswillbeyourgithub** [comment]
Just fyi, the very fine folks over at [openedai-speech](https://github.com/matatonic/openedai-speech/pull/29) are currently implementing automatic language detection for piper. Meaning it will transparently switch voices that are trained on the detected language in the sentence etc. It's in the dev branch

**Comment by salocharly** [comment]
any solution for non technical people?? can find any info on how to setup azure. but i want to run local.

**Reply by chrishart0** [reply]
Check my repo, you just need to spin up their docker container and fiddle with some configs. If you get stuck just let me know. What GPU do you have? It does change depending on if you've got an Nvidia GPU with CUDA or not.

https://github.com/chrishart0/open-webui-nvidia-cuda-setup

**Comment by maximegmd** [comment]
Does anyone have an alternative to `openedai-speech` with a more permissive license?

**Reply by rsxdalv** [reply]
Chatterbox has multiple OpenAI compatible APIs on Github

**Comment by deifius** [comment]
kyutai labs has released a voice interface featuring full duplex spoken conversation capabilities.  https://github.com/kyutai-labs/moshi

**Comment by d4v3y0rk** [comment]
I just started using this project as a direct replacement for edge-tts https://github.com/remsky/Kokoro-FastAPI
it works perfectly and the output audio is amazing and all local.

**Reply by TJW65** [reply]
@d4v3y0rk Is there something I'm missing with this?
Using the following settings, it appears to be loading indefinitely for me. 

Text-to-Speech Engine: "http://host.docker.internal/8880/v1"
api_key: "not-needed"
TTS Voice: "af_sky"
TTS Model: "kokoro"
also tried: http://localhost:8880/v1

The Gradio interface on port 7860 works fine.
I would be grateful if you'd share your steps. :)

**Reply by d4v3y0rk** [reply]
![image](https://github.com/user-attachments/assets/0dda723f-4e8c-4393-b11e-ff1d6de63e39)
These are my settings in the ui. 
I cloned the repo and ran `docker compose up -d --build`
The only thing I did change in the docker compose file was I had something already running on port 7860 so I changes it to 7000 for the UI but I have not even looked at the UI for it.

**Reply by d4v3y0rk** [reply]
This is my command for running open-webui 
`docker run -d --gpus all --network=host -v open-webui:/app/backend/data -e OLLAMA_BASE_URL=http://127.0.0.1:11434 --name open-webui --restart always ghcr.io/open-webui/open-webui:cuda`

**Reply by d4v3y0rk** [reply]
I opened a PR with docs for using this setup. [Here](https://github.com/open-webui/docs/pull/354)

**Reply by TJW65** [reply]
I found the problem.
For whatever reason I needed to set the PCs IP adress. Using "localhost" didn't work. 
Never seen that before. Works flawlessly now though.

**Reply by d4v3y0rk** [reply]
Awesome, it works pretty great IMHO. The voices sound really good and it is all local which is my favorite part.

**Reply by silentoplayz** [reply]
Thanks @d4v3y0rk! I've followed along and got Kokoro TTS Fast API up and running and integrated as the Text-to-Speech engine for Open WebUI within minutes! All the voice options available with Kokoro are MUCH clearer than what I considered to be the "best" voice option for me from openedai-speech, which was the `alloy` voice.

**Reply by thiswillbeyourgithub** [reply]
Really nice BUT english only though.

**Comment by digitalassassins** [comment]
For anyone using Windows who wants a more naturally speaking voice without having to install a separate LLM & Docker. 
I only got 3 installed options in the drop-down for the 3 default legacy Windows Voices (Sound Like Robots).
I went to Narrator settings and installed the natural voices, but they would not show in Open Web UI.
I found and installed the NaturalVoiceSAPIAdapter and now all the Natural Voices, including the ones included in Edge, now show up in the list in Open Web UI and sound better. - https://github.com/gexgd0419/NaturalVoiceSAPIAdapter
I have on my mobile a WebView APK for Open Web UI tunnelling through using Tailscale.
Kokoro.js on my PC sounded great, but soon as I switched to it on Android it would crash the app or Chrome out.
The default Android TTS sounds great. If you change your sound settings on Windows it changes it on the logged in account on Android as well. Maybe it should allow you to have different audio settings on different devices. For instance, save the TTS sound settings in Cookies instead of in the program itself.
Now I have another problem. Both devices sound great, but they read out the asterisk over and over. and it is very annoying.
Ive tried writing a Python script to strip the asterisk out of the stream and the final output text. Stripping it out of the stream stopped the Markup from being displayed correctly (Worked but was harder to read) and stripping it out of the final output text did nothing. The TTS still read out the asterisks.
Any body know how to strip asterisk characters out of the text before it's sent to the TTS for synthesis?

**Comment by rsxdalv** [comment]
Currently I provide Chatterbox and Kokoro via TTS WebUI's Open-AI style API.
My hope is to get a cleaner 'snap-in' integration but I cannot find any information on how to integrate with the "audio" settings (other than reading the whole codebase).
Adding more models that are already in TTS WebUI is not too difficult, but requires interest from community.
